File: examples/matlab/pulse_ep_client.m
% PULSE_EP_CLIENT  Minimal REST client for pulse-ep (R2020a+).
%
%   This file defines four utility functions that talk to a running
%   pulse-ep server over its JSON REST API. They use only built-in
%   MATLAB facilities (webread / webwrite) — no Toolbox required.
%
%   token   = pe_login(baseURL, username, password)
%   studies = pe_list_studies(baseURL, token)
%   maps    = pe_list_maps(baseURL, token, studyId)
%   mesh    = pe_get_mesh(baseURL, token, mapId, scalarName, distance)
%   areas   = pe_areas_per_interval(baseURL, token, mapId, intervals, scalarName, distance)
%
%   See pulse_ep_demo.m and areas_per_interval.m for runnable examples.

function pulse_ep_client  %#ok<*STOUT>
    error('pulse_ep_client.m only defines helper functions; call them directly.');
end

% -----------------------------------------------------------------------
function token = pe_login(baseURL, username, password)
%PE_LOGIN  Exchange username/password for a JWT bearer token.
    body = jsonencode(struct('username', username, 'password', password));
    opts = weboptions( ...
        'MediaType',     'application/json', ...
        'ContentType',   'json', ...
        'RequestMethod', 'POST', ...
        'Timeout',       30);
    resp  = webwrite(strcat(baseURL, "/login_user"), body, opts);
    token = resp.access_token;
end

% -----------------------------------------------------------------------
function studies = pe_list_studies(baseURL, token)
%PE_LIST_STUDIES  Return a struct array with fields id, study_name.
    opts    = weboptions('HeaderFields', ...
              {'Authorization', char(strcat("Bearer ", token))}, ...
              'ContentType', 'json', 'Timeout', 30);
    studies = webread(strcat(baseURL, "/list_studies"), opts);
end

% -----------------------------------------------------------------------
function maps = pe_list_maps(baseURL, token, studyId)
%PE_LIST_MAPS  List EP maps in a study.
    opts = weboptions('HeaderFields', ...
           {'Authorization', char(strcat("Bearer ", token))}, ...
           'ContentType', 'json', 'Timeout', 30);
    maps = webread( ...
        sprintf("%s/list_epmaps_in_study/%d", baseURL, studyId), opts);
end

% -----------------------------------------------------------------------
function mesh = pe_get_mesh(baseURL, token, mapId, scalarName, distance)
%PE_GET_MESH  Fetch one map's mesh and per-vertex scalars.
%
%   Returns a struct with fields:
%     mesh.vertices       [Nv x 3] double
%     mesh.faces          [Nt x 3] int32     (1-based for MATLAB)
%     mesh.scalars        [Nv x 1] double    (NaN outside `distance`)
%     mesh.normalized     [Nv x 1] double
%     mesh.points         [Np x 3] double    catheter coordinates
%     mesh.point_scalars  [Np x 1] double
    if nargin < 4 || isempty(scalarName); scalarName = "act";    end
    if nargin < 5 || isempty(distance);   distance   = 5.0;       end

    opts = weboptions('HeaderFields', ...
           {'Authorization', char(strcat("Bearer ", token))}, ...
           'ContentType', 'json', 'Timeout', 60);
    data = webread( ...
        strcat(baseURL, "/get_mesh_data"), opts, ...
        'map_id', mapId, 'scalar_name', scalarName, 'distance', distance);

    md = data.mesh_data;
    pd = data.point_data;

    mesh.vertices   = local_to_matrix(md.vertices, 3);
    mesh.faces      = int32(local_to_matrix(md.faces, 3)) + 1;       % 0→1-indexed
    mesh.scalars    = local_to_column(md.scalar_data);
    mesh.normalized = local_to_column(md.normalized_scalar_data);

    if isfield(pd, 'coordinates') && ~isempty(pd.coordinates)
        mesh.points = local_to_matrix(pd.coordinates, 3);
    else
        mesh.points = zeros(0, 3);
    end
    if isfield(pd, 'scalar_data') && ~isempty(pd.scalar_data)
        mesh.point_scalars = local_to_column(pd.scalar_data);
    else
        mesh.point_scalars = zeros(0, 1);
    end
end

% -----------------------------------------------------------------------
function areas = pe_areas_per_interval(baseURL, token, mapId, intervals, ...
                                       scalarName, distance)
%PE_AREAS_PER_INTERVAL  Call /calculate_areas_for_intervals.
%
%   intervals : N-by-2 numeric matrix of [lo hi] bins (e.g. [50 60; 60 70; ...]).
%   areas     : Nx1 double, surface area in cm^2 per interval (NaN for empty).
    if nargin < 5 || isempty(scalarName); scalarName = "act"; end
    if nargin < 6 || isempty(distance);   distance   = 5.0;   end

    % MATLAB serialises [N x 2] as nested JSON arrays — exactly what the
    % Flask endpoint expects.
    payload = struct( ...
        'map_id',      mapId, ...
        'scalar_name', scalarName, ...
        'distance',    distance, ...
        'intervals',   intervals);
    body = jsonencode(payload);

    opts = weboptions( ...
        'HeaderFields', {'Authorization', char(strcat("Bearer ", token)); ...
                         'Content-Type',  'application/json'}, ...
        'MediaType',     'application/json', ...
        'ContentType',   'json', ...
        'RequestMethod', 'POST', ...
        'Timeout',       60);
    resp = webwrite(strcat(baseURL, "/calculate_areas_for_intervals"), body, opts);

    raw = resp.areas;
    if iscell(raw)
        areas = cellfun(@local_or_nan, raw);
        areas = areas(:);
    else
        areas = double(raw(:));
    end
end

% -----------------------------------------------------------------------
function M = local_to_matrix(cellOfRows, ncols)
%LOCAL_TO_MATRIX  Convert a cell-of-rows (from JSON) to a numeric matrix.
    if iscell(cellOfRows)
        rows = cellfun(@(r) reshape(double(r), 1, ncols), ...
                       cellOfRows, 'UniformOutput', false);
        M = vertcat(rows{:});
    else
        % If MATLAB already simplified it to a 2-D numeric array.
        M = double(cellOfRows);
    end
end

% -----------------------------------------------------------------------
function v = local_to_column(values)
%LOCAL_TO_COLUMN  Convert a cell or array of (numeric|null) to a column.
    if iscell(values)
        v = cellfun(@(x) local_or_nan(x), values);
        v = v(:);
    else
        v = double(values(:));
    end
end

function x = local_or_nan(v)
    if isempty(v); x = NaN; else; x = double(v); end
end

File: examples/r/pulse_ep_client.R
# pulse_ep_client.R — Minimal REST client for pulse-ep, built on httr2.
#
# Usage:
#   source("pulse_ep_client.R")
#   token  <- pe_login()                       # reads env vars
#   studies <- pe_list_studies(token)
#   maps    <- pe_list_maps(token, study_id = studies$id[1])
#   mesh    <- pe_get_mesh(token, map_id = maps$id[1])

suppressPackageStartupMessages({
  library(httr2)
  library(jsonlite)
})

#' Resolve the pulse-ep base URL, honoring the PULSE_EP_BASE_URL env var.
pe_base_url <- function(base_url = NULL) {
  if (!is.null(base_url) && nzchar(base_url)) return(base_url)
  env <- Sys.getenv("PULSE_EP_BASE_URL", unset = "")
  if (!nzchar(env)) {
    stop(
      "pulse-ep base URL is not set. Pass `base_url=` or export ",
      "PULSE_EP_BASE_URL.",
      call. = FALSE
    )
  }
  env
}

#' Exchange username/password for a JWT access token.
pe_login <- function(username = Sys.getenv("PULSE_EP_USERNAME"),
                     password = Sys.getenv("PULSE_EP_PASSWORD"),
                     base_url = NULL) {
  if (!nzchar(username) || !nzchar(password)) {
    stop(
      "PULSE_EP_USERNAME / PULSE_EP_PASSWORD are not set in the env.",
      call. = FALSE
    )
  }
  resp <- request(pe_base_url(base_url)) |>
    req_url_path("/login_user") |>
    req_method("POST") |>
    req_headers("Content-Type" = "application/json") |>
    req_body_raw(toJSON(list(username = username, password = password),
                       auto_unbox = TRUE)) |>
    req_perform()
  resp_body_json(resp)$access_token
}

#' Internal: GET helper that adds the Bearer token.
.pe_get <- function(token, path, query = list(), base_url = NULL) {
  req <- request(pe_base_url(base_url)) |>
    req_url_path(path) |>
    req_headers(Authorization = paste("Bearer", token))
  if (length(query) > 0) req <- req_url_query(req, !!!query)
  resp_body_json(req_perform(req), simplifyVector = TRUE)
}

#' List all studies. Returns a data.frame with columns `id` and `study_name`.
pe_list_studies <- function(token, base_url = NULL) {
  data <- .pe_get(token, "/list_studies", base_url = base_url)
  if (length(data) == 0) return(data.frame(id = integer(), study_name = character()))
  as.data.frame(data, stringsAsFactors = FALSE)
}

#' List the EP maps belonging to one study.
pe_list_maps <- function(token, study_id, base_url = NULL) {
  data <- .pe_get(token, paste0("/list_epmaps_in_study/", study_id),
                  base_url = base_url)
  if (length(data) == 0) return(data.frame())
  as.data.frame(data, stringsAsFactors = FALSE)
}

#' Call /calculate_areas_for_intervals — surface area per score bin.
#'
#' `intervals` is a list of length-2 numeric vectors, e.g.
#'   list(c(50, 60), c(60, 70), c(70, 80), c(80, 90), c(90, 100)).
#' Returns a numeric vector of areas in cm^2, NA for empty bins.
pe_areas_per_interval <- function(token, map_id, intervals,
                                  scalar_name = NULL, distance = 5.0,
                                  base_url = NULL) {
  payload <- list(
    map_id      = map_id,
    scalar_name = scalar_name,
    distance    = distance,
    intervals   = intervals
  )
  resp <- request(pe_base_url(base_url)) |>
    req_url_path("/calculate_areas_for_intervals") |>
    req_method("POST") |>
    req_headers(Authorization = paste("Bearer", token),
                "Content-Type" = "application/json") |>
    req_body_raw(toJSON(payload, auto_unbox = TRUE)) |>
    req_perform()
  data <- resp_body_json(resp, simplifyVector = TRUE)
  as.numeric(unlist(data$areas))
}

#' Fetch the mesh (vertices, triangles, scalars) for one map.
#' Returns a list with `vertices`, `faces`, `scalars`, `normalized`,
#' `points` and `point_scalars`.
`%||%` <- function(a, b) if (is.null(a)) b else a

#' Quantities a map carries: name, kind and unit, plus the primary one.
#' Clients should read the choice from here rather than assuming a field name.
pe_map_scalars <- function(token, map_id) {
  req <- request(paste0(pe_base_url(), "/epmaps/", map_id, "/scalars")) |>
    req_headers(Authorization = paste("Bearer", token))
  resp <- req_perform(req)
  out <- resp_body_json(resp, simplifyVector = TRUE)
  out
}

pe_get_mesh <- function(token, map_id, scalar_name = NULL,
                        distance = 5.0, base_url = NULL) {
  data <- .pe_get(
    token, "/get_mesh_data",
    query = list(
      map_id = map_id, scalar_name = scalar_name, distance = distance
    ),
    base_url = base_url
  )
  md <- data$mesh_data
  pd <- data$point_data
  list(
    vertices       = matrix(unlist(md$vertices), ncol = 3, byrow = TRUE),
    faces          = matrix(unlist(md$faces),    ncol = 3, byrow = TRUE) + 1L,  # R is 1-indexed
    scalars        = as.numeric(md$scalar_data),
    normalized     = as.numeric(md$normalized_scalar_data),
    # An empty JSON array arrives as an empty *list*, not NULL, so a
    # !is.null() guard lets it through and matrix() then fails on the NULL
    # that unlist() returns. A map may legitimately have no acquisition
    # points: an EnSiteX mesh carrying only per-vertex fields has none.
    points         = if (length(pd$coordinates) > 0)
                       matrix(unlist(pd$coordinates), ncol = 3, byrow = TRUE)
                     else matrix(numeric(), ncol = 3),
    point_scalars  = as.numeric(pd$scalar_data)
  )
}
